src/osm/GeometryHandler.cpp

```diff
@@ -61,137 +79,402 @@ osm2rdf::osm::GeometryHandler<W>::GeometryHandler(
-         "\"part\":\"" +
-         std::string(part) +
-         "\","
-         "\"check\":\"" +
-         std::string(check) +
-         "\","
-         "\"outer_id\":" +
-         std::to_string(outerId) +
-         ","
-         "\"outer_type\":\"" +
-         std::string(outerType) +
-         "\","
-         "\"inner_id\":" +
-         std::to_string(innerId) +
-         ","
-         "\"inner_type\":\"" +
-         std::string(innerType) +
-         "\","
-         "\"duration_ns\":" +
-         std::to_string(durationNS.count()) +
-         ","
-         "\"result\":" +
-         (result ? "true" : "false") +
-         ""
-         "},\n";
-}
-
-// ____________________________________________________________________________
-template <typename W>
-void osm2rdf::osm::GeometryHandler<W>::area(const osm2rdf::osm::Area& area) {
-  auto geom = area.geom();
-  if (_config.simplifyGeometries > 0) {
-    geom = simplifyGeometry(geom);
+GeometryHandler<W>::~GeometryHandler() = default;
+
+// ____________________________________________________________________________
+template <typename W>
+void GeometryHandler<W>::relation(const Relation& rel) {
+  // careful, tags is returned via copy (why?)
+  const auto& tags = rel.tags();
+  auto typeTag = tags.find("type");
+  if (typeTag != tags.end() &&
+      (typeTag->second == "boundary" || typeTag->second == "multipolygon")) {
+    for (const auto& member : rel.members()) {
+      if (member.type() == RelationMemberType::WAY &&
+          (member.role() == "outer" || member.role() == "inner")) {
+#pragma omp critical
+        {
+          _areaBorderWaysIndex[member.id()].push_back(
+              {rel.id(), member.role() == "inner"});
+          std::sort(_areaBorderWaysIndex[member.id()].begin(),
+                    _areaBorderWaysIndex[member.id()].end(), MemberRelCmp());
+        }
+      }
+    }
+  }
+}
+
+// ____________________________________________________________________________
+template <typename W>
+void GeometryHandler<W>::area(const Area& area) {
+  // inner simplified geom, empty by default
+  osm2rdf::geometry::Area innerGeom;
+
+  // outer simplified geom, empty by default
+  osm2rdf::geometry::Area outerGeom;
+
+  // the convex hull of the geometry
+  osm2rdf::geometry::Polygon convexHull;
+
+  const auto& geom = simplifyGeometry(area.geom());
+
+  size_t totPoints = 0;
+  const size_t MIN_CUTOUT_POINTS = 10000;
+
+  for (const auto& poly : geom) {
+    totPoints += poly.outer().size();
+  }
+
+  if (!_config.dontUseInnerOuterGeoms && (area.hasName() || !area.fromWay())) {
+    innerGeom = simplifiedArea(geom, true);
+    outerGeom = simplifiedArea(geom, false);
+  }
+
+  if (!geom.empty()) {
+    boost::geometry::convex_hull(geom, convexHull);
+    if (convexHull.outer().size() >= geom.front().outer().size() / 2) {
+      // if we haven't gained anything significant in terms of number of anchor
+      // points, discard
+      convexHull = osm2rdf::geometry::Polygon();
+    }
```

Comment:
Just a comment for the future:
For oriented-bounding-boxes the convex hull is required if [rotating calipers](https://en.wikipedia.org/wiki/Rotating_calipers) is used. This could mean that everything but Nodes will need this geometry later.